### PR TITLE
Spec supporting trigger context ID for aggregate reporting

### LINKED ISSUE #974: Further reducing aggregatable report delays
Recent feedback ([meeting notes](https://github.com/WICG/attribution-reporting-api/blob/main/meetings/2023-08-21-minutes.md)) indicates that even the new modest report delays for aggregatable reports are not ideal, and we should explore the instant-style approaches outlined in https://github.com/WICG/attribution-reporting-api/blob/main/report_verification.md#could-we-just-tag-reports-with-a-trigger_id-instead-of-using-anonymous-tokens. Here is what a privacy neutral enhancement to the API could look like that enables this use-case:

Trigger registration could come along with an optional top-level parameter that ad techs can set: `trigger_event_id`[^1]. If this parameter is set, a few things happen:
- Aggregatable reports will be generated for this registration unconditionally, even if it was not preceded by a source event. In that case any generated report will act like the [existing null aggregatable reports](https://wicg.github.io/attribution-reporting-api/#generate-null-reports) i.e. they will not affect any aggregate computation.
- Aggregatable reports associated with this registration will be delivered immediately after the browser receives the registration.
- Aggregatable reports associated with this registration will come bearing a new piece of cleartext information in its body which passes the `trigger_event_id`.

By default this will result in a single report per registration. However, if `aggregatable_source_registration_time` is ”include”, it will require generating 30 reports for every trigger registration. Given that 30 reports is non-ideal from a system health perspective, we may want to consider disallowing this combination, and only support instant reports if the `aggregatable_source_registration_time` is “exclude” (feedback welcome though!). The purpose of these “unconditional” reports is to ensure that the output of the trigger registration does not leak any cross-site information (i.e. whether the user had previously seen an ad or not).

cc @thegreatfatzby , @bmayd, and @dmdabbs who voiced support for reducing delays all the way to 0 in our Aug 21 call, especially in relation to CPA-based billing.

Note: this proposal may greatly increase (up to 20x[^2]) the number of reports you receive and need to query the aggregation service with, which can impact aggregation query latency. Here is a derivation to find out how it will impact you:

- Let `true_rate` be the fraction of registrations that lead to a true report (i.e. have a previous matching source)
- With no noise, the number of reports is `true_rate * num_triggers`
- Currently, with the 5% flip probability the number of reports is `((1 - true_rate) * .05 + true_rate) * num_triggers = (.95 * true_rate + .05) * num_triggers`
- With this proposal, the number of reports is `num_triggers`. Thus, adopting this proposal would result in `(1 / (.95 * true_rate + .05))` times more total reports than the status quo (note that this smoothly interpolates between 0x and 20x at the extremes).

Note with `true_rate = .2`, this gives you ~4x times more total reports than the status quo.

[^1]: Keen observers will note the similarities between this and the `context_id` param in https://github.com/patcg-individual-drafts/private-aggregation-api/blob/main/report_verification.md#shared-storage
[^2]: Assuming aggregatable_source_registration_time is “exclude”.

### PR DESCRIPTION
**Pull request by linnan-github**
Fixes #974 

Explainer: https://github.com/WICG/attribution-reporting-api/pull/1113

***
<a href="https://pr-preview.s3.amazonaws.com/linnan-github/conversion-measurement-api/pull/1114.html" title="Last updated on Nov 21, 2023, 1:49 PM UTC (909ebee)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/1114/ed340d6...linnan-github:909ebee.html" title="Last updated on Nov 21, 2023, 1:49 PM UTC (909ebee)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1172,6 +1176,10 @@ Its value is «[=source type/navigation=] → 8, [=source type/event=] → 2».
 controls the maximum [=map/size=] of a [=trigger spec map=] for an
 [=attribution source=]. Its value is 32.
 
+<dfn>Max length per trigger context ID</dfn> is a positive integer that controls
+the maximum [=string/length=] of an [=attribution trigger=]'s [=attribution trigger/trigger context ID=],
+and an [=aggregatable report=]'s [=aggregatable report/trigger context ID=]. Its value is 64.
```
Comment (apasel422) [inline index.bs:1181]: Similar to https://github.com/WICG/attribution-reporting-api/pull/1113/files#r1400607818 I'd rather not explicitly say that this constant limits the length of the *report* field, but rather just the input field.

File: index.bs [hunk 2]
```diff
@@ -2425,6 +2435,13 @@ and a [=moment=] |triggerTime|:
     1. If |value|["`aggregatable_source_registration_time`"] is not an [=aggregatable source registration time configuration=],
         return null.
     1. Set |aggregatableSourceRegTimeConfig| to |value|["`aggregatable_source_registration_time`"].
+1. Let |triggerContextID| be null.
+1. If |value|["`trigger_context_id`"] [=map/exists=]:
+    1. If |value|["`trigger_context_id`"] is not a [=string=], return null.
+    1. If |value|["`trigger_context_id`"] is empty or its [=string/length=] is greater than the [=max length per trigger context ID=],
```
Comment (apasel422) [inline index.bs:2441]:
```suggestion
    1. If |value|["`trigger_context_id`"]'s [=string/length=] is 0 or is greater than the [=max length per trigger context ID=],
```

File: index.bs [hunk 3]
```diff
@@ -3172,13 +3196,17 @@ To <dfn>generate null reports</dfn> given an [=attribution trigger=] |trigger| a
 
 1. Let |nullReports| be a new [=list/is empty|empty=] [=list=].
 1. If |trigger|'s [=attribution trigger/aggregatable source registration time configuration=] is "<code>[=aggregatable source registration time configuration/exclude=]</code>":
+    1. Let |randomizedNullReportRate| be [=randomized null report rate excluding source registration time=].
+    1. If |trigger|'s [=attribution trigger|trigger context ID=] is not null, set
```
Comment (apasel422) [inline index.bs:3200]:
```suggestion
    1. If |trigger|'s [=attribution trigger/trigger context ID=] is not null, set
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 1 above)
Comment (linnan-github) [reply index.bs]: Updated, thanks.

**Review by apasel422 — APPROVED** [review]

**Pull request merged**